### PR TITLE
Rewrite MouseTracker's tracking and notifying algorithm

### LINKED ISSUE #40000: When transitioning from MouseRegions onEnter callback is executed before onExit
## Steps to Reproduce

Consider the following snippet:

```dart
void main() {
  debugDefaultTargetPlatformOverride = TargetPlatform.fuchsia;
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Column(
          children: <Widget>[
            MouseRegion(
              onEnter: (_) => print('entered 1'),
              onExit: (_) => print('left 1'),
              child: Container(
                color: Colors.blue,
                height: 100,
                width: 100,
                child: Center(
                  child: Text('1'),
                ),
              ),
            ),
            MouseRegion(
              onEnter: (_) => print('entered 2'),
              onExit: (_) => print('left 2'),
              child: Container(
                color: Colors.green,
                height: 100,
                width: 100,
                child: Center(
                  child: Text('2'),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


```

When the mouse enters Region 1 and then moves over to Region 2 we see printed:

```
flutter: entered 1
flutter: entered 2
flutter: left 1
```

which is counter intuitive and I would expect that the onExit callback for Region 1 to be called before the onEnter callback for Region 2 is, furthermore this behavior makes it hard to implement certain functionality where keeping track of the currently hovered widget is necessary.

## Logs

```
[✓] Flutter (Channel unknown, v1.8.5-pre.86, on Mac OS X 10.14.6 18G87, locale en-US)
    • Flutter version 1.8.5-pre.86 at /Users/xxxx/repos/xxxx/flutter
    • Framework revision f1ea485018 (4 weeks ago), 2019-08-08 14:48:07 -0700
    • Engine revision f200ee13aa
    • Dart version 2.5.0 (build 2.5.0-dev.1.0 f29f41f1a5)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/setup/#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, set ANDROID_HOME to that location.
      You may also want to add it to your PATH environment variable.


[✓] Xcode - develop for iOS and macOS (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • CocoaPods version 1.7.5

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/setup/#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2019.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 39.0.5
    • Dart plugin version 192.6527

[✓] Connected device (1 available)
    • macOS • macOS • darwin-x64 • Mac OS X 10.14.6 18G87

! Doctor found issues in 2 categories.
```

### PR DESCRIPTION
**Pull request by dkwingsmt**
## Description

This PR redesigns the algorithm with which `MouseTracker` keeps track of states and dispatches events. It keeps the functionality of current methods. While the new algorithm changes the order of dispatched events, it should be considered a bug fix since the current behavior is not well defined and has been reported as problematic (#40000).

The major differences of the new design include:
- The relationship between annotations and devices are now tracked per-device instead of a mixture of per-device and per-annotation. This not only brings a cleaner structure, but also has the potential benefit of removing the need of keeping track of all annotations on screen (which isn't included in this PR due to being breaking).
- The dispatching algorithm has been redesigned with a well-defined order between enter, exit, and hover events, as proposed in https://github.com/flutter/flutter/issues/41420.

This PR also adds tests. Previous tests are kept but some are revised, include
- 'receives and processes mouse hover events'
  - _split into_ 'should detect enter, hover, and exit from Added, Hover, and Removed events'
  - _and_ 'should correctly handle multiple devices'
- 'detects exit when annotated layer no longer hit'
  - _changed to_ 'should detect enter or exit when annotations are attached or detached on the pointer'
- 'detects exit when mouse goes away'
  - _merged into_ 'should detect enter, hover, and exit from Added, Hover, and Removed events'
- 'handles mouse down and move'
  - _changed to_ 'should not handle non-hover events'

This PR also fixes a few bugs, including `MouseTrackerAnnotation.toString` was unreadable.

## Related Issues

- https://github.com/flutter/flutter/issues/41420 describes the new algorithm
- Fixes https://github.com/flutter/flutter/issues/40000

## Tests

I added the following tests:
- `MouseTracker` correctly reacts to mouse state changes
- `MouseTracker` correctly reacts to annotation state changes
- `MouseTracker` dispatches callbacks in the correct order
- `MouseTrackerAnnotation` has the same string representation

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 1]
```diff
@@ -104,60 +129,77 @@ class MouseTracker extends ChangeNotifier {
...
+  //
+  // If a device disconnects, the device will be moved to
+  // `_disconnectedMouseState`.
+  final Map<int, _MouseState> _mouseStates = <int, _MouseState>{};
+
+  // Tracks the state of the disconnected mouse device before its exit callback
+  // is called.
+  //
+  // A device is moved from `_mouseStates` on its disconnecting event, and
+  // will be permanantly removed after its exit callback.
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: I don't understand why you need this field at all. Can you not just mark a mouse as disconnected in its _MouseState and then remove it from _mouseStates after sendMouseNotifications has been called in _removeMouseDevice?

File: packages/flutter/test/gestures/mouse_tracking_test.dart [hunk 2]
```diff
@@ -16,217 +16,576 @@ import '../flutter_test_alternative.dart';
...
-    final ui.PointerDataPacket packet5 = ui.PointerDataPacket(data: <ui.PointerData>[
-      _pointerData(PointerChange.hover, const Offset(1.0, 401.0), device: 1),
-    ]);
-    RendererBinding.instance.mouseTracker.attachAnnotation(annotation);
-    RendererBinding.instance.mouseTracker.sendMouseNotifications(<int>{0});
-    isInHitRegionOne = true;
-    ui.window.onPointerDataPacket(packet1);
+  test('MouseTrackerAnnotation has correct toString', () {
+    final MouseTrackerAnnotation annotation1 =
+      MouseTrackerAnnotation(onEnter: (_) {}, onExit: (_) {}, onHover: (_) {});
```
Comment (goderbauer) [inline packages/flutter/test/gestures/mouse_tracking_test.dart]:
```suggestion
    final MouseTrackerAnnotation annotation1 = MouseTrackerAnnotation(
      onEnter: (_) {}, 
      onExit: (_) {}, 
      onHover: (_) {},
    );
```

File: packages/flutter/test/gestures/mouse_tracking_test.dart [hunk 3]
```diff
@@ -16,217 +16,576 @@ import '../flutter_test_alternative.dart';
...
+    expect(
+      annotation1.toString(),
+      equals('MouseTrackerAnnotation#${shortHash(annotation1)}(callbacks: enter hover exit)'),
+    );
+
+    const MouseTrackerAnnotation annotation2 =
+      MouseTrackerAnnotation();
```
Comment (goderbauer) [inline packages/flutter/test/gestures/mouse_tracking_test.dart]: keep on previous line.

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 4]
```diff
@@ -209,131 +248,143 @@ class MouseTracker extends ChangeNotifier {
...
-    }
-
-    void exitAllDevices(_TrackedAnnotation trackedAnnotation) {
-      if (trackedAnnotation.activeDevices.isNotEmpty) {
-        final Set<int> deviceIds = trackedAnnotation.activeDevices.toSet();
-        for (int deviceId in deviceIds) {
-          exitAnnotation(trackedAnnotation, deviceId);
+    final Iterable<_MouseState> targetDevices = deviceIds != null ?
+      deviceIds.map(_findMouseState) :
+      () sync* {
```
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: I was about to but worried that creating a new array is slower than iterator. What's your opinion?

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 5]
```diff
@@ -209,131 +248,143 @@ class MouseTracker extends ChangeNotifier {
...
-              if (trackedAnnotation.annotation?.onExit != null) {
-                trackedAnnotation.annotation.onExit(PointerExitEvent.fromMouseEvent(lastEvent));
-              }
-              trackedAnnotation.activeDevices.remove(deviceId);
-            }
+      // Send hover events in reverse visual order.
+      // No solid reasons have been brough up on the order between the hover
+      // events, except for keeping it aligned with enter events for simplicity.
+      if (mostRecentEvent is PointerHoverEvent) {
+        final Iterable<MouseTrackerAnnotation> hoveringAnnotations = nextAnnotations
```
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: It's too long to fit into a single line. What about splitting at the assignment, such as
```dart
        final Iterable<MouseTrackerAnnotation> hoveringAnnotations =
          nextAnnotations.toList().reversed;
```

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 6]
```diff
@@ -209,131 +248,143 @@ class MouseTracker extends ChangeNotifier {
...
+      // Order is important for mouse event callbacks. The `findAnnotations`
+      // returns annotations in the visual order from front to back. We call
+      // it the "visual order", and the opposite one "reverse visual order".
+      // The algorithm here is explained in
+      // https://github.com/flutter/flutter/issues/41420
+
+      // The annotations that contains this device in the coming frame in
+      // visual order.
+      final Set<MouseTrackerAnnotation> nextAnnotations = thisDeviceIsConnected
+        ? annotationFinder(mouseState.mostRecentEvent.position).toSet()
```
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: Good point. I've changed to `Set.from`, which is documented as "The created [Set] is a [LinkedHashSet]". What do you think?

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 7]
```diff
@@ -48,49 +48,74 @@ class MouseTrackerAnnotation {
...
-  /// This is used to detect layers that used to have the mouse pointer inside
-  /// them, but now no longer do (to facilitate exit notification).
-  Set<int> activeDevices = <int>{};
-}
-
-/// Describes a function that finds an annotation given an offset in logical
-/// coordinates.
+/// Signature for searching for [MouseTrackerAnnotation]s from the given offset.
 ///
 /// It is used by the [MouseTracker] to fetch annotations for the mouse
```
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: I added document at `MouseTracker.annotationFinder`, since the signature itself doesn't convey this idea.

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 8]
```diff
@@ -209,131 +248,143 @@ class MouseTracker extends ChangeNotifier {
...
   ///
-  /// This is called from a post-frame callback when the layer tree has been
-  /// updated, right after rendering the frame.
+  /// This is called synchronously in most cases, except for when a new
+  /// annotation is attached, which is called in a post frame callback.
   ///
   /// This function is only public to allow for proper testing of the
   /// MouseTracker. Do not call in other contexts.
   @visibleForTesting
   void sendMouseNotifications(Iterable<int> deviceIds) {
```
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: Yes, it is the plan. It's not done in this PR because changing the signature is breaking. I have a PR (https://github.com/flutter/flutter/pull/41879) that makes this method private, after which we can do whatever change that makes it pretty.

File: packages/flutter/lib/src/gestures/mouse_tracking.dart (hunk 1 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: It was indeed the initial design, but then I discovered that there can only be one disconnected mouse state at a given time, and even that is temporary that only lasts during 2 lines within `_removeMouseDevice`. Therefore it doesn't really worth assigning another field for all mouse states.
I plan to move it as a parameter of `sendMouseNotifications` after the method becomes private, so that `_disconnectedMouseState` can be removed completely, but for now I think this is the easiest plan.

**Review by dkwingsmt — COMMENTED** [review]
All comments have been taken care of.

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 9]
```diff
@@ -48,49 +49,80 @@ class MouseTrackerAnnotation {
...
+//
+// It is used by [MouseTracker] to compute which callbacks should be triggered
+// by each event.
+class _MouseState {
+  _MouseState({
+    @required PointerEvent mostRecentEvent,
+  }) : assert(mostRecentEvent != null),
+       _mostRecentEvent = mostRecentEvent;
+
+  // The list of annotations that contains this device during the last frame.
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: maybe document that this is in a certain order (to make clear why this is explicitly a LinkedHashSet).

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 10]
```diff
@@ -48,49 +49,80 @@ class MouseTrackerAnnotation {
...
+  LinkedHashSet<MouseTrackerAnnotation> lastAnnotations =
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: nit: Leave this on one line to improve readability (Flutter's style guide does not have a hard 80 char cutoff limit, you can go over to make things more readable)

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 11]
```diff
@@ -48,49 +49,80 @@ class MouseTrackerAnnotation {
...
+
+/// Maintains the relationship between mouse devices and
+/// [MouseTrackerAnnotation]s, and notifies interested callbacks of the changes
+/// thereof.
 ///
 /// This class is a [ChangeNotifier] that notifies its listeners if the value of
 /// [mouseIsConnected] changes.
 ///
-/// Owned by the [RendererBinding] class.
+/// This class is a global singleton and owned by the [RendererBinding] class.
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: nit: The class itself is not a global singelton, I can easily instantiate more than one instance. Maybe rephrase this to say that RendererBinding owns a global singelton of this class?

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 12]
```diff
@@ -48,49 +49,80 @@ class MouseTrackerAnnotation {
...
 class MouseTracker extends ChangeNotifier {
   /// Creates a mouse tracker to keep track of mouse locations.
   ///
+  /// The first parameter is a [PointerRouter], which [MouseTracker] will
+  /// subscribe to and receive events from. Usually it is global singleton
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: ... it is the global ...

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 13]
```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
...
   }
 
-  // The pointer router that the mouse tracker listens to for events.
+  /// Find annotations at a given offset in global logical coordinate space
+  /// in visual order from front to back.
+  ///
+  /// [MouseTracker] uses this callback to know which annotations are affected
+  /// by each device.
+  ///
+  /// The order between the returned annotations is important since it
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: The order of the returned annotations....

This should also say what the order needs to be.

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 14]
```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
...
-        annotation.onExit(PointerExitEvent.fromMouseEvent(event));
+  void _removeMouseDevice(int device, PointerEvent event) {
+    final bool wasConnected = mouseIsConnected;
+    assert(_mouseStates.containsKey(device));
+    final _MouseState disconnectedMouseState = _mouseStates.remove(device);
+    disconnectedMouseState.mostRecentEvent = event;
+    // Schedule a check to exit annotations that used to contain this pointer.
+    _checkDeviceUpdates(
+      device: device,
+      disconnectedMouseState: disconnectedMouseState,
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: Why do you need to pass this in separately (seems like it complicates the logic in this method as well)? Could you add a boolean "disconnected" to MouseState and flip it to true here instead of removing the mouse state from _mouseStates. And then after _checkDeviceUpdates comes back, you remove it from the list?

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 15]
```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
...
+    final int device = event.device;
+    if (event is PointerAddedEvent) {
+      _addMouseDevice(device, event);
+    } else if (event is PointerRemovedEvent) {
+      _removeMouseDevice(device, event);
+    } else if (event is PointerHoverEvent) {
+      final _MouseState mouseState = _guaranteeMouseState(device, event);
+      final PointerEvent previousEvent = mouseState.mostRecentEvent;
+      mouseState.mostRecentEvent = event;
+      if (previousEvent is PointerAddedEvent || previousEvent.position != event.position) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: Why do we need to send a hover event if previousEvent is PointerAddedEvent, but the position hasn't changed?

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 16]
```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
...
+  // but this mouse will be hovering no annotations.
+  void _checkDeviceUpdates({
+    int device,
+    _MouseState disconnectedMouseState,
+  }) {
+    final _MouseState mouseState = disconnectedMouseState ?? _mouseStates[device];
+    final bool thisDeviceIsConnected = mouseState != disconnectedMouseState;
+    assert(mouseState != null);
+
+    final LinkedHashSet<MouseTrackerAnnotation> nextAnnotations =
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: This block is awkwardly formatted. The lines with the ? and : need to be indented more, I think?

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 17]
```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
...
   }
 
-  // Handler for events coming from the PointerRouter.
-  void _handleEvent(PointerEvent event) {
-    if (event.kind != PointerDeviceKind.mouse) {
-      return;
+  // Dispatch callbacks related to a device after all necessary information
+  // has been collected.
+  //
+  // This function should not not change the provided states, and should not
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: not not?

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 18]
```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
...
+
+    // The `nextAnnotations` is annotations that contains this device in the
+    // coming frame in visual order.
+    // Order is preserved becuase [Set.from] creates a [LinkedHashSet], which
+    // keeps the order.
+
+    final PointerEvent mostRecentEvent = currentState.mostRecentEvent;
+    // The `lastAnnotations` is annotations that contains this device in the
+    // previous frame in visual order.
+    final LinkedHashSet<MouseTrackerAnnotation> lastAnnotations
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: nit: one line.

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 19]
```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
...
-  void _handleEvent(PointerEvent event) {
-    if (event.kind != PointerDeviceKind.mouse) {
-      return;
+  // Dispatch callbacks related to a device after all necessary information
+  // has been collected.
+  //
+  // This function should not not change the provided states, and should not
+  // access information that is not provided in parameters (hence being
+  // static).
+  static void _dispatchDeviceCallbacks({
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: Maybe move this method after _checkDeviceUpdates. It's only called from there and you kinda have to read that one first to understand what's going on.

File: packages/flutter/test/gestures/mouse_tracking_test.dart [hunk 20]
```diff
@@ -16,211 +16,579 @@ import '../flutter_test_alternative.dart';
...
-    ]);
-    isInHitRegionOne = true;
-    RendererBinding.instance.mouseTracker.attachAnnotation(annotation);
-    ui.window.onPointerDataPacket(packet1);
-    ui.window.onPointerDataPacket(packet2);
-    expect(events, _equalToEventsOnCriticalFields(<PointerEvent>[
-      const PointerEnterEvent(position: Offset(0.0, 0.0), delta: Offset(0.0, 0.0)),
-      const PointerHoverEvent(position: Offset(0.0, 0.0), delta: Offset(0.0, 0.0)),
-      const PointerHoverEvent(position: Offset(1.0, 101.0), delta: Offset(1.0, 101.0)),
+  test('should trigger callbacks between parents and children in correctly order', () {
```
Comment (gspencergoog) [inline packages/flutter/test/gestures/mouse_tracking_test.dart]:
```suggestion
  test('should trigger callbacks between parents and children in correct order', () {
```

File: packages/flutter/lib/src/gestures/mouse_tracking.dart (hunk 14 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: I'm concerned that keeping disconnected states in `_mouseStates` will affect `mouseIsConnected`. If the callback called by `_checkDeviceUpdates` checks `mouseIsConnected`, it'll get a `true` even when the last mouse has been disconnected.

File: packages/flutter/lib/src/gestures/mouse_tracking.dart (hunk 11 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: I changed it to:
```dart
/// An instance of [MouseTracker] is owned by the global singleton of
/// [RendererBinding].
```

File: packages/flutter/lib/src/gestures/mouse_tracking.dart (hunk 14 above)
Comment (goderbauer) [reply packages/flutter/lib/src/gestures/mouse_tracking.dart]: I see. Makes sense.

**Comment by dkwingsmt** [comment]
I moved `_dispatchDeviceCallbacks` to the end of the private method section, among some smaller doc fixes.

**Review by goderbauer — APPROVED** [review]
LGTM
